pennylane/qnode.py

```diff
@@ -605,19 +605,20 @@ def __call__(self, *args, **kwargs):
             # construct the tape
             self.construct(args, kwargs)
 
-        # Execute the tape.
         # If the observable contains a Hamiltonian and the device does not
-        # support Hamiltonians, split tape into multiple tapes and execute
-        # sequentially. In future, this logic should be moved to the device
+        # support Hamiltonians, or if the simulation uses finite shots,
+        # split tape into multiple tapes of diagonalizable known observables.
+        # In future, this logic should be moved to the device
         # to allow for more efficient batch execution.
-        supports_hamiltonian = self.device.supports_observable("Hamiltonian")
+        does_not_support_hamiltonian = not self.device.supports_observable("Hamiltonian")
+        finite_shots = self.device.shots is not None
         hamiltonian_in_obs = "Hamiltonian" in [obs.name for obs in self.qtape.observables]
-        if hamiltonian_in_obs and not supports_hamiltonian:
+        if hamiltonian_in_obs and (does_not_support_hamiltonian or finite_shots):

```

Comment:
Until the device can preprocess tapes before entering the blackbox (planned!) we always split Hamiltonians if finite shots are asked for - else we have a situation that the device gets `expval(H)` and `shots=1000`, and it can only raise an error because splitting the tape at that stage would not be tracked by autodiff.